File: AdventOfCode2020/AdventOfCode2020_/Types/DefaultDictionary.vb
Public Class DefaultDictionary(Of TKey, TValue As New)
	Inherits Dictionary(Of TKey, TValue)

	Default Public Overloads Property Item(ByVal key As TKey) As TValue
		Get
			Dim val As TValue

			If Not TryGetValue(key, val) Then
				val = New TValue()
				Add(key, val)
			End If

			Return val
		End Get
		Set(ByVal value As TValue)
			MyBase.Item(key) = value
		End Set
	End Property

End Class

File: AdventOfCode2021/AdventOfCode2021/Types/DefaultDictionary.cs
using System.Collections.Generic;

namespace AdventOfCode2021.Types
{
	internal class DefaultDictionary<TKey, TValue> : Dictionary<TKey, TValue> where TValue : struct
	{
		private readonly TValue Default = default;

		public DefaultDictionary()
		{ }

		public DefaultDictionary(TValue defaultValue) => Default = defaultValue;

		public new TValue this[TKey key]
		{
			get
			{
				if (!TryGetValue(key, out TValue value))
				{
					value = Default;
					Add(key, value);
				}
				return value;
			}
			set => base[key] = value;
		}
	}
}